contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,208 @@
+# Background
+
+ As part of increasing security of a cluster, we are planning to limit the ability
+of a given Kubelet (in general: node), to be able to read only resources associated
+with it. Those resources, in particular means: secrets, configmaps &
+persistentvolumeclaims. This is needed to avoid situation when compromising node
+de facto means compromising a cluster. For more details & discussions see
+https://github.com/kubernetes/kubernetes/issues/40476.
+ 
+ However, by some extension to this effort, we would like to improve scalability
+of the system, by significantly reducing amount of api calls coming from kubelets.
+As of now, to avoid situation that kubelet is watching all secrets/configmaps/...
+in the system, it is not using watch. Instead of that, it is retrieving indidual pods,
+by sending individual GET requests. However, it is sending those requests periodically
+to enable automatic updates of mounted secrets/configmaps/... In large clusters,
+this is generating huge unnecessary load, as this load in principle should be
+watch-based. We would like to address this together with solving the authorization issue.
+
+# Proposal
+
+ In this proposal, I'm not focusing on how exactly security should be done - I'm just
+sketching very high level approach and exact authorization mechanism should be discussed
+separately.
+
+ At the high level, what we want to achieve is to enable LIST and WATCH requests to
+support filtering of "objects only attached to pods bound to a given node". We obviously
+want to be able to authorize other types of requests (in particular GETs), so that the
+design has to be consistent with that.
+
+ To solve this, I propose to introduce a new filtering mechanism (next to label selector
+and field selector): ```node selector ``` (we probably need better name though). Its
+semantic will be to filter only objects that are attached to pods bound to a given node
+and it will be supported for some predefined set of object types.
+
+# Detailed design
+
+ We will introduce the following ```node selector ``` filtering mechanism:
```

Comment:
So there's a sig-auth discussion about allowing selective retrieval of secrets for the purposes of using them.  In that model you'd have to say up front who you are and access them via a subresource (potentially so that you can get redirected to a third party system).  So we're already considering decoupling secrets strongly, which makes watches impossible.

Part of the reason for direct access is because it keeps the core simple.  Yes, we could do all sorts of crazy filtering, and watch for changing ACL rules, and match those up on the fly to access.  The question is - why bother, if we can put in place a reasonably performant "lazy" access?

https://github.com/kubernetes/kubernetes/issues/40403 had gone into some of the options here, but the sheer simplicity of having only two types of filter on LIST WATCH:

1. field/label
2. can you start watching X

means that the scale complexity is at least tractable - you can effectively detect when ACL changes and break the watch (which we need to start doing at some point in the future).  But when you bring in linking and relationships, you have to do a much more complex join, and the potential security risks are much higher if you get the join ordering wrong.

That said, if we can define a maximum window for watch breaking, we could in theory do the ACL check on a single item watch every X minutes (at the window for maximum time open) and look at allowing bulk WATCH (watch these keys).  Not sure how computationally tractable that is, but it separates WATCH from ACL still, which I think is a net win.